EntryNotFoundException thrown in federated registries when using custom input/output configuration

**Describe the bug**  
When running with `federatedRegistries-1.0`, it is possible to get an EntryNotFoundException when defining a non-identifier type property for the federated registries input/output mapping. This exception can occur in any of the `com.ibm.ws.security.wim.registry.util.*Bridge` classes, but the key is it originates from a `BridgeUtils.getEntityByIdentifier` call.

Below is an example stack:

```
com.ibm.websphere.security.EntryNotFoundException: CWIML1010E: The user registry operation could not be completed. The uniqueId = null and uniqueName = null attributes of the identifier object are either not valid or not defined in the back-end repository.
	at com.ibm.ws.security.registry.internal.UserRegistryWrapper.getGroupsForUser(UserRegistryWrapper.java:248)
	at web.UserRegistryServlet.handleMethodRequest(UserRegistryServlet.java:140)
	at web.UserRegistryServlet.doGet(UserRegistryServlet.java:174)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.service(ServletWrapper.java:1258)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:746)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:443)
	at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1227)
	at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1011)
	at com.ibm.ws.webcontainer.servlet.CacheServletWrapper.handleRequest(CacheServletWrapper.java:75)
	at com.ibm.ws.webcontainer.WebContainer.handleRequest(WebContainer.java:938)
	at com.ibm.ws.webcontainer.osgi.DynamicVirtualHost$2.run(DynamicVirtualHost.java:281)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink$TaskWrapper.run(HttpDispatcherLink.java:1246)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.wrapHandlerAndExecute(HttpDispatcherLink.java:468)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.ready(HttpDispatcherLink.java:427)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleDiscrimination(HttpInboundLink.java:567)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleNewRequest(HttpInboundLink.java:501)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.processRequest(HttpInboundLink.java:361)
	at com.ibm.ws.http.channel.internal.inbound.HttpICLReadCallback.complete(HttpICLReadCallback.java:70)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.requestComplete(WorkQueueManager.java:514)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.attemptIO(WorkQueueManager.java:584)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.workerRun(WorkQueueManager.java:968)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager$Worker.run(WorkQueueManager.java:1057)
	at com.ibm.ws.threading.internal.ExecutorServiceImpl$RunnableWrapper.run(ExecutorServiceImpl.java:245)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:836)
Caused by: com.ibm.ws.security.registry.EntryNotFoundException: CWIML1010E: The user registry operation could not be completed. The uniqueId = null and uniqueName = null attributes of the identifier object are either not valid or not defined in the back-end repository.
	at com.ibm.ws.security.wim.registry.util.BridgeUtils.handleExceptions(BridgeUtils.java:587)
	at com.ibm.ws.security.wim.registry.util.MembershipBridge.getGroupsForUser(MembershipBridge.java:260)
	at com.ibm.ws.security.wim.registry.WIMUserRegistry.getGroupsForUser(WIMUserRegistry.java:484)
	at com.ibm.ws.security.registry.internal.UserRegistryWrapper.getGroupsForUser(UserRegistryWrapper.java:244)
	... 27 more
Caused by: com.ibm.wsspi.security.wim.exception.InvalidIdentifierException: CWIML1010E: The user registry operation could not be completed. The uniqueId = null and uniqueName = null attributes of the identifier object are either not valid or not defined in the back-end repository.
	at com.ibm.ws.security.wim.ProfileManager.getImpl(ProfileManager.java:421)
	at com.ibm.ws.security.wim.ProfileManager.genericProfileManagerMethod(ProfileManager.java:263)
	at com.ibm.ws.security.wim.ProfileManager.get(ProfileManager.java:207)
	at com.ibm.ws.security.wim.VMMService.get(VMMService.java:211)
	at com.ibm.ws.security.wim.registry.util.BridgeUtils.getEntityByIdentifier(BridgeUtils.java:468)
	at com.ibm.ws.security.wim.registry.util.MembershipBridge.getGroupsForUser(MembershipBridge.java:124)
	... 29 more
```

**Steps to Reproduce**  
Configure an LDAP registry. Configure the userSecurityNameMapping input property to a non-identifier type property, such as mail below.

```
    <ldapRegistry realm="LdapRealm"... >
        ...
    </ldapRegistry>

    <federatedRepository>
      <primaryRealm name="FederatedRealm">
	<participatingBaseEntry name="dc=example,dc=com"/>
	<userSecurityNameMapping inputProperty="mail" outputProperty="mail"/>
      </primaryRealm>
    </federatedRepository>
```

Call the getGroupsForUsers UserRegistry method and you will receive the EntryNotFoundException. Again, this would apply to all of the `com.ibm.ws.security.wim.registry.util.*Bridge` classes. The key to reproducing is setting the input mapping for the returned value for that specific method to one that is not an identifier type property (uniqueName, uniqueId, externalName, externalId).

**Expected behavior**  
The method should not throw an EntryNotFoundException. The bridge class should NOT call the WIM GET method, but should instead call the SEARCH method. The GET method can only be utilized using an identifier property (uniqueName, uniqueId, externalName, externalId).

**Diagnostic information:**  
 - OpenLiberty Version: ALL
 - Affected feature(s): federatedRegistry-1.0
 - Java Version: Any
 - server.xml configuration: see above

**Additional context**  
None.

--- a/dev/com.ibm.websphere.security_fat.1/test-applications/WebsphereUserRegistry/src/web/UserRegistryServlet.java
+++ b/dev/com.ibm.websphere.security_fat.1/test-applications/WebsphereUserRegistry/src/web/UserRegistryServlet.java
@@ -1,10 +1,10 @@
 /*******************************************************************************
- * Copyright (c) 2011, 2020 IBM Corporation and others.
+ * Copyright (c) 2011, 2023 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License 2.0
  * which accompanies this distribution, and is available at
  * http://www.eclipse.org/legal/epl-2.0/
- * 
+ *
  * SPDX-License-Identifier: EPL-2.0
  *
  * Contributors:
@@ -29,7 +29,6 @@ import com.ibm.websphere.security.PasswordCheckFailedException;
 import com.ibm.websphere.security.Result;
 import com.ibm.websphere.security.UserRegistry;
 import com.ibm.websphere.security.WSSecurityException;
-import com.ibm.ws.security.registry.RegistryException;
 import com.ibm.wsspi.security.registry.RegistryHelper;
 
 /**
@@ -46,7 +45,7 @@ public class UserRegistryServlet extends HttpServlet {
      * @param realmName realm name.
      *
      * @return UserRegistry instance, or null if can't find service.
-     * @throws RegistryException
+     * @throws CustomRegistryException
      */
     private UserRegistry getCurrentUserRegistry(String realmName) throws CustomRegistryException {
         try {
@@ -80,51 +79,70 @@ public class UserRegistryServlet extends HttpServlet {
         UserRegistry ur = getCurrentUserRegistry(realmName);
         try {
             if ("getRealm".equals(method)) {
+                pw.println("Calling: " + method + "()");
                 response = ur.getRealm();
             } else if ("checkPassword".equals(method)) {
                 String userSecurityName = req.getParameter("userSecurityName");
                 String password = req.getParameter("password");
+                pw.println("Calling: " + method + "(" + userSecurityName + ", xxxxxxx)");
                 response = String.valueOf(ur.checkPassword(userSecurityName, password));
             } else if ("mapCertificate".equals(method)) {
                 response = "mapCertificate is not supported via the servlet";
             } else if ("isValidUser".equals(method)) {
                 String userSecurityName = req.getParameter("userSecurityName");
+                pw.println("Calling: " + method + "(" + userSecurityName + ")");
                 response = String.valueOf(ur.isValidUser(userSecurityName));
             } else if ("getUsers".equals(method)) {
                 String pattern = req.getParameter("pattern");
                 int limit = Integer.valueOf(req.getParameter("limit"));
+                pw.println("Calling: " + method + "(" + pattern + ", " + limit + ")");
                 response = convertResultToString(ur.getUsers(pattern, limit));
             } else if ("getUserDisplayName".equals(method)) {
                 String userSecurityName = req.getParameter("userSecurityName");
+                pw.println("Calling: " + method + "(" + userSecurityName + ")");
                 response = ur.getUserDisplayName(userSecurityName);
             } else if ("getUniqueUserId".equals(method)) {
                 String userSecurityName = req.getParameter("userSecurityName");
+                pw.println("Calling: " + method + "(" + userSecurityName + ")");
                 response = ur.getUniqueUserId(userSecurityName);
             } else if ("getUserSecurityName".equals(method)) {
                 String uniqueUserId = req.getParameter("uniqueUserId");
+                pw.println("Calling: " + method + "(" + uniqueUserId + ")");
                 response = ur.getUserSecurityName(uniqueUserId);
             } else if ("isValidGroup".equals(method)) {
                 String groupSecurityName = req.getParameter("groupSecurityName");
+                pw.println("Calling: " + method + "(" + groupSecurityName + ")");
                 response = String.valueOf(ur.isValidGroup(groupSecurityName));
             } else if ("getGroups".equals(method)) {
                 String pattern = req.getParameter("pattern");
                 int limit = Integer.valueOf(req.getParameter("limit"));
+                pw.println("Calling: " + method + "(" + pattern + ", " + limit + ")");
                 response = convertResultToString(ur.getGroups(pattern, limit));
             } else if ("getGroupDisplayName".equals(method)) {
                 String groupSecurityName = req.getParameter("groupSecurityName");
+                pw.println("Calling: " + method + "(" + groupSecurityName + ")");
                 response = ur.getGroupDisplayName(groupSecurityName);
             } else if ("getUniqueGroupId".equals(method)) {
                 String groupSecurityName = req.getParameter("groupSecurityName");
+                pw.println("Calling: " + method + "(" + groupSecurityName + ")");
                 response = ur.getUniqueGroupId(groupSecurityName);
             } else if ("getGroupSecurityName".equals(method)) {
                 String uniqueGroupId = req.getParameter("uniqueGroupId");
+                pw.println("Calling: " + method + "(" + uniqueGroupId + ")");
                 response = ur.getGroupSecurityName(uniqueGroupId);
             } else if ("getUniqueGroupIdsForUser".equals(method)) {
                 String uniqueUserId = req.getParameter("uniqueUserId");
+                pw.println("Calling: " + method + "(" + uniqueUserId + ")");
                 response = String.valueOf(ur.getUniqueGroupIds(uniqueUserId));
             } else if ("getGroupsForUser".equals(method)) {
                 String userSecurityName = req.getParameter("userSecurityName");
+                pw.println("Calling: " + method + "(" + userSecurityName + ")");
                 response = String.valueOf(ur.getGroupsForUser(userSecurityName));
+            } else if ("getUsersForGroup".equals(method)) {
+                String groupSecurityName = req.getParameter("groupSecurityName");
+                int limit = Integer.valueOf(req.getParameter("limit"));
+                pw.println("Calling: " + method + "(" + groupSecurityName + ", " + limit + ")");
+                response = convertResultToString(ur.getUsersForGroup(groupSecurityName, limit));
             } else {
                 pw.println("Usage: url?method=name&paramName=paramValue&...");
             }

--- a/dev/com.ibm.ws.security.wim.registry/src/com/ibm/ws/security/wim/registry/util/BridgeUtils.java
+++ b/dev/com.ibm.ws.security.wim.registry/src/com/ibm/ws/security/wim/registry/util/BridgeUtils.java
@@ -1,10 +1,10 @@
 /*******************************************************************************
- * Copyright (c) 2012, 2020 IBM Corporation and others.
+ * Copyright (c) 2012, 2023 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License 2.0
  * which accompanies this distribution, and is available at
  * http://www.eclipse.org/legal/epl-2.0/
- * 
+ *
  * SPDX-License-Identifier: EPL-2.0
  *
  * Contributors:
@@ -438,8 +438,20 @@ public class BridgeUtils implements WIMUserRegistryDefines {
 
         Root returnValue = null;
 
-        // check if the input attribute is an identifier type
-        if (isInputAttrIdentifier || hasFederatedRegistry) {
+        /*
+         * If we have at least one federated UserRegistry and we are using the default PrincipalName or CN input attribute names,
+         * we will translate them into a UniqueName input attribute name.
+         */
+        if (hasFederatedRegistry && (SchemaConstants.PROP_PRINCIPAL_NAME.equalsIgnoreCase(inputAttrName) || "CN".equalsIgnoreCase(inputAttrName))) {
+            inputAttrName = SchemaConstants.PROP_UNIQUE_NAME;
+            isInputAttrIdentifier = true;
+        }
+
+        /*
+         * If the input attribute is NOT an identifier attribute we cannot make a GET call to WIM as it requires
+         * an identifier attribute.
+         */
+        if (isInputAttrIdentifier) {
             if (!isOutputAttrIdentifier) {
                 mapUtils.createPropertyControlDataObject(root, outputAttrName);
             }
@@ -452,10 +464,7 @@ public class BridgeUtils implements WIMUserRegistryDefines {
             }
             if (entity != null) {
                 IdentifierType idfType = new IdentifierType();
-                if (SchemaConstants.PROP_PRINCIPAL_NAME.equalsIgnoreCase(inputAttrName) || "CN".equalsIgnoreCase(inputAttrName))
-                    idfType.set(SchemaConstants.PROP_UNIQUE_NAME, inputAttrValue);
-                else
-                    idfType.set(inputAttrName, inputAttrValue);
+                idfType.set(inputAttrName, inputAttrValue);
                 entity.setIdentifier(idfType);
             }
 

